### PR TITLE
#1402 MetricsUtil improvement

### PR DESCRIPTION
**Pull request by xnull**
## Overview

Description:
MetricsUtil improvement

Why should this be merged:
In the prev version, we had more than one MetricRegistry in several places, the current version provides only one MetricsRegistry (in MetricsUtil)

Related issue(s) (if applicable): #1402 

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

**Comment by xnull** [comment]
done, I have deprecated the old behavior

File: runtime/src/main/java/org/corfudb/util/MetricsUtils.java [hunk 1]
```diff
@@ -46,20 +40,27 @@ private MetricsUtils() {
     private static final String PROPERTY_LOG_INTERVAL = "corfu.metrics.log.interval";
     private static final String PROPERTY_METRICS_COLLECTION = "corfu.metrics.collection";
 
+    private static final long metricsCsvInterval;
+    private static final long metricsLogInterval;
+    private static final String metricsCsvFolder;
+    @Getter
+    private static final boolean metricsCollectionEnabled;
+    private static boolean metricsCsvReportingEnabled;
+    private static final boolean metricsJmxReportingEnabled;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/util/MetricsUtils.java]: CRITICAL Rename this constant name to match the regular expression '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00115)

File: runtime/src/main/java/org/corfudb/util/MetricsUtils.java [hunk 2]
```diff
@@ -46,20 +40,27 @@ private MetricsUtils() {
...
+    private static final boolean metricsJvmCollectionEnabled;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/util/MetricsUtils.java]: CRITICAL Rename this constant name to match the regular expression '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00115)

File: runtime/src/main/java/org/corfudb/util/MetricsUtils.java [hunk 3]
```diff
@@ -46,20 +40,27 @@ private MetricsUtils() {
...
+    private static final boolean metricsSlf4jReportingEnabled;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/util/MetricsUtils.java]: CRITICAL Rename this constant name to match the regular expression '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00115)

File: runtime/src/main/java/org/corfudb/util/MetricsUtils.java [hunk 4]
```diff
@@ -46,20 +40,27 @@ private MetricsUtils() {
...
+    private static final String mpTrigger = "filter-trigger"; // internal use only
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/util/MetricsUtils.java]: CRITICAL Rename this constant name to match the regular expression '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00115)

File: runtime/src/main/java/org/corfudb/util/MetricsUtils.java [hunk 5]
```diff
@@ -46,20 +40,27 @@ private MetricsUtils() {
     private static final String PROPERTY_LOG_INTERVAL = "corfu.metrics.log.interval";
     private static final String PROPERTY_METRICS_COLLECTION = "corfu.metrics.collection";
 
+    private static final long metricsCsvInterval;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/util/MetricsUtils.java]: CRITICAL Rename this constant name to match the regular expression '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00115)

File: runtime/src/main/java/org/corfudb/util/MetricsUtils.java [hunk 6]
```diff
@@ -46,20 +40,27 @@ private MetricsUtils() {
...
+    private static final long metricsLogInterval;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/util/MetricsUtils.java]: CRITICAL Rename this constant name to match the regular expression '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00115)

File: runtime/src/main/java/org/corfudb/util/MetricsUtils.java [hunk 7]
```diff
@@ -46,20 +40,27 @@ private MetricsUtils() {
...
+    private static final String metricsCsvFolder;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/util/MetricsUtils.java]: CRITICAL Rename this constant name to match the regular expression '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00115)

File: runtime/src/main/java/org/corfudb/util/MetricsUtils.java [hunk 8]
```diff
@@ -46,20 +40,27 @@ private MetricsUtils() {
...
+    @Getter
+    private static final boolean metricsCollectionEnabled;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/util/MetricsUtils.java]: CRITICAL Rename this constant name to match the regular expression '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00115)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 9]
```diff
@@ -370,20 +370,22 @@
...
     static {
-        synchronized (defaultMetrics) {
-            MetricsUtils.metricsReportingSetup(defaultMetrics);
-        }
+        MetricsUtils.metricsReportingSetup();
     }
 
+    @Deprecated
+    @Getter
+    private static final MetricRegistry defaultMetrics = MetricsUtils.metrics;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 10]
```diff
@@ -370,20 +370,22 @@
     @Getter
     private volatile boolean isShutdown = false;
 
-    @Getter
-    private static MetricRegistry defaultMetrics = new MetricRegistry();
+    @Deprecated
     @Getter
     @Setter
-    private MetricRegistry metrics = new MetricRegistry();
+    private MetricRegistry metrics = defaultMetrics;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 16 issues
* ![CRITICAL][CRITICAL] 8 critical
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 4 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [LogUnitServer.java#L3](https://github.com/CorfuDB/CorfuDB/blob/9138ee912dd6ceedb6d7a4ca2f43a82ab39e2949/infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java#L3): Remove this unused import 'com.codahale.metrics.MetricRegistry'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)
1. ![MINOR][MINOR] [ServerContext.java#L3](https://github.com/CorfuDB/CorfuDB/blob/9138ee912dd6ceedb6d7a4ca2f43a82ab39e2949/infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java#L3): Remove this unused import 'com.codahale.metrics.MetricRegistry'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)
1. ![MINOR][MINOR] [AddressSpaceView.java#L4](https://github.com/CorfuDB/CorfuDB/blob/9138ee912dd6ceedb6d7a4ca2f43a82ab39e2949/runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java#L4): Remove this unused import 'com.codahale.metrics.Gauge'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)
1. ![MINOR][MINOR] [AddressSpaceView.java#L5](https://github.com/CorfuDB/CorfuDB/blob/9138ee912dd6ceedb6d7a4ca2f43a82ab39e2949/runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java#L5): Remove this unused import 'com.codahale.metrics.MetricRegistry'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by xnull** [comment]
My assumption is to use one registry but different namespaces for different project/modules. The final destination of all metrics must be an analytics/ visualization tool like grafana, so it seems to me that a good idea to have only one metric registry or at least have all of them in MetricsUtils. What do you think?

**Comment by xnull** [comment]
@Maithem I wrote a test for cvs metrics report also I will squash commit during the merge to master

**Comment by StrikingThirteen** [comment]
Having different namespaces doesn't address differently treating the collection and reporting of different metrics. Also Grafana will be only another reporter to which the dropwizard is going to report (through Graphite). Note that we already have a host of other reporters which must be kept as they are required for their own purpose (e.g. logging/ JMX / etc.). 

To step back, besides the clean-up, what is the mentioned improvement that this commit tries to do?

**Comment by xnull** [comment]
This PR won't be merged. After discussion with @StrikingThirteen we decided not to touch current version but design a better version of the metrics tool. Will discuss it soon in meetings.

**Pull request closed without merging**